### ISSUE TITLE
Wrong stream created if flush is not done before Z_FINISH.

### ISSUE THREAD

**Issue by alexey-milovidov** [zlib-ng/zlib-ng#494]
Test case:

```
#include <unistd.h>
#include <vector>
#include <stdexcept>
#include <zlib.h>

#pragma GCC diagnostic ignored "-Wold-style-cast"


int main(int, char **)
{
    std::vector<unsigned char> in(1048576);
    std::vector<unsigned char> out(1048576);

    // Slightly incorrect but works if stdin is file, not a pipe.
    ssize_t in_size = read(STDIN_FILENO, in.data(), 1048576);
    if (in_size < 0)
        throw std::runtime_error("Cannot read");
    in.resize(in_size);

    z_stream zstr{};
    if (Z_OK != deflateInit2(&zstr, 1, Z_DEFLATED, 15 + 16, 8, Z_DEFAULT_STRATEGY))
        throw std::runtime_error("Cannot deflateInit2");

    zstr.next_in = in.data();
    zstr.avail_in = in.size();
    zstr.next_out = out.data();
    zstr.avail_out = out.size();

    while (zstr.avail_in > 0)
        if (Z_OK != deflate(&zstr, Z_NO_FLUSH))
            throw std::runtime_error("Cannot deflate");

    while (true)
    {
        int rc = deflate(&zstr, Z_FINISH);

        if (rc == Z_STREAM_END)
            break;

        if (rc != Z_OK)
            throw std::runtime_error("Cannot finish deflate");
    }

    deflateEnd(&zstr);

    if (ssize_t(zstr.total_out) != write(STDOUT_FILENO, out.data(), zstr.total_out))
        throw std::runtime_error("Cannot write");

    return 0;
}
```

```
$ seq 0 34536 > data.tsv
$ ./main < data.tsv > data.tsv.gz
$ gzip -d < data.tsv.gz | tail

gzip: stdin: unexpected end of file
34527
34528
34529
34530
34531
34532
34533
34534
34535
34536
```

It's reproducing under the following conditions:
- level 1 is used;
- there is no separate call to `deflate` with any flush mode before call with Z_FINISH;
- upper bound of sequence of numbers is from 34536 to 34619;

**Comment by mtl1979** [comment]
Because you are telling it not to flush, it will overflow the internal data structures happily... last block of input data will show correctly as it did fit in the deflate window. I suspect same happens with higher compression levels, but require larger sequence of numbers.

**Comment by alexey-milovidov** [comment]
> Because you are telling it not to flush, it will overflow the internal data structures happily... 

It will stop reproducing after 34620.

> I suspect same happens with higher compression levels, but require larger sequence of numbers.

With higher compression levels it doesn't reproduce for all numbers below 100 000 and also for numbers around 1 000 000, so I don't think it reproduces at all.

I have read the documentation of `deflate` in zlib.h and have not found that I have to flush periodically. And I don't see any separate flushes in examples available...

**Comment by mtl1979** [comment]
Test ```if (Z_OK != deflate(&zstr, Z_NO_FLUSH))``` is insufficient according to the zlib documentation... it doesn't check for output buffer overflow as in avail_out == 0.

**Comment by alexey-milovidov** [comment]
This is just an example to reproduce the issue. In this example, the size of output buffer is big enough, so the condition `avail_out == 0` never happens.

You can see the real code here: https://github.com/ClickHouse/ClickHouse/blob/master/dbms/src/IO/ZlibDeflatingWriteBuffer.cpp#L57

**Comment by mtl1979** [comment]
Level 1 use deflate_quick which is not included in the original documentation... Only relevant internal structure is pending_buf which in normal cases should not overflow. I only see one line in deflate_quick that mentions Z_NO_FLUSH.

**Comment by nmoinvaz** [comment]
I'm not able to reproduce this. I used this code which is similar but can be used in _example.c_:

```
void test_deflate1_no_flush(unsigned char *compr, size_t comprLen, unsigned char *uncompr, size_t uncomprLen)
{
    int32_t err;
    PREFIX3(stream) c_stream;

    c_stream.zalloc = zalloc;
    c_stream.zfree = zfree;
    c_stream.opaque = (voidpf)0;

    err = PREFIX(deflateInit2)(&c_stream, 1, Z_DEFLATED, 15 + 16, 8, Z_DEFAULT_STRATEGY);
    CHECK_ERR(err, "deflateInit2");

    c_stream.next_in = uncompr;
    c_stream.avail_in = uncomprLen;
    c_stream.next_out = compr;
    c_stream.avail_out = uncomprLen;

    /*
    while (c_stream.avail_in > 0) {
        err = PREFIX(deflate)(&c_stream, Z_NO_FLUSH);
        CHECK_ERR(err, "deflate");
    }
    */

    for (;;) {
        err = PREFIX(deflate)(&c_stream, Z_FINISH);
        if (err == Z_STREAM_END) break;
        CHECK_ERR(err, "deflate");
    }
    err = PREFIX(deflateEnd)(&c_stream);
    CHECK_ERR(err, "deflateEnd");

    FILE *f = fopen("extest.gz", "wb");
    fwrite(compr, 1, c_stream.total_out, f);
    fclose(f);

    return;
}
```

And passed `compr` and `uncompr` 1mb buffers. The gz file it produced extracted successfully at least on my Windows machine. I'm not sure what the "sequence of numbers" is referring to in the original issue description. If there is not enough data to write in `avail_out` then `deflate` will return `Z_BUF_ERROR` AFAIK.

**Comment by nmoinvaz** [comment]
This is most likely resolved with #620.

**Comment by nmoinvaz** [comment]
This should be fixed now that #620 is merged.
